### ISSUE TITLE
[jvm-packages] Distributed computing with H2O

### ISSUE THREAD

**Issue by mdymczyk** [dmlc/xgboost#2340]
Following #2032 's example I'm creating this issue as I'm implementing distributed XGBoost integration in [H2O](https://github.com/h2oai/h2o-3). I've already done most of the work following your Spark/Flink implementation but I have several questions, all of them assume a distributed training setting:

1) when creating a `DMatrix` should it be after `Rabit.init()` or not? Should the env passed to Rabit be the distributed one or can I just pass an empty map. Spark/Flink are doing it after `init`, Dask is doing it before.

2) Is running first `XGBoost.train` with 0 rounds to make the initial booster, shutting down Rabit, reinitializing Rabit for subsequent `booster.update(dmatrix, iter)` supported? Seems like so but I want to be sure. I do this as I want to make my own stats after each iteration.

3) Can I, in the same JVM, first call `Rabit.init()` with an empty map (not distributed) to do some initial `train()` and then `Rabit.init()` with an env map gotten from RabitTracker to do distributed training? Or will the initial init override something?

Asking because the following scenario seems to not work:

Driver:
```
RabitTracker rt = new RabitTracker(nWorkers);
rt.start(0);
// pass rt.getEnv() to workers
// start distribtued job
rt.waitFor(0); // finishes just fine - gets the shutdown signals!
```

On workers:
```
Rabit.init(emptyMap);
ml.dmlc.xgboost4j.java.XGBoost.train(trainMat, gpuParams, 1, watches, null, null); // checks if there's a GPU on the machine, in real code there's a try/catch and some param setting based on the result
Rabit.shutdown();

rabitEnv.put("DMLC_TASK_ID", nodeIndex); //env from the driver, local to each worker
Rabit.init(rabitEnv);
DMatrix trainMat = XGBoost.convertFrametoDMatrix(h2oFrame); // local part of the h2oFrame on this node
booster = ml.dmlc.xgboost4j.java.XGBoost.train(trainMat, params, 50, watches, null, null);
Rabit.shutdown();
```

Logs clearly show:
```
Tracker started, with env={DMLC_NUM_SERVER=0, DMLC_TRACKER_URI=172.16.2.199, DMLC_TRACKER_PORT=9091, DMLC_NUM_WORKER=2}
DEBUG Recieve start signal from 172.16.2.199; assign rank 0
DEBUG Recieve start signal from 172.16.2.204; assign rank 1
INFO @tracker All of 2 nodes getting started
DEBUG Recieve shutdown signal from 0
DEBUG Recieve shutdown signal from 1
INFO @tracker All nodes finishes job
``` 

Furthermore `Rabit.allReduce(new float[]{1}, Rabit.OpType.MAX);` works properly returning the right sum (2) but the booster model doesn't seem to work, getting between ~50-80% accuracy, whereas if I run a single node setup on the same data/params I'm getting ~95%. It looks to me as if both models are training in separation and don't exchange info.

####

This much for now, sorry for the lengthy "issue" but couldn't find the answers in the docs and going through the code is taking me a bit of time.

**Comment by CodingCat** [comment]
> when creating a DMatrix should it be after Rabit.init() or not? Should the env passed to Rabit be the distributed one or can I just pass an empty map. Spark/Flink are doing it after init, Dask is doing it before.

I think it should be after....I have the impression that DMatrix is (unnecessarily) calling some rabit stuffs...but forget about the details....

> Is running first XGBoost.train with 0 rounds to make the initial booster, shutting down Rabit, reinitializing Rabit for subsequent booster.update(dmatrix, iter) supported? Seems like so but I want to be sure. I do this as I want to make my own stats after each iteration.

I do not think it is supported, you shutdown rabit means you tell the tracker, I am gone, and since we cannot "update" tree by growing, we cannot train in this way

> Can I, in the same JVM, first call Rabit.init() with an empty map (not distributed) to do some initial train() and then Rabit.init() with an env map gotten from RabitTracker to do distributed training? Or will the initial init override something?

XGBoost does not have support for (full) model update, so I do not think you can do this

**Comment by mdymczyk** [comment]
@CodingCat thanks for the answer! So I did a bit of testing and a basic version is already up and running!

1) didn't see much difference doing DMatrix creating before or after `Rabit.init`so I guess both are ok

2) tried out something along the lines of

```
Rabit.init(distributedEnvMap);
booster.update(dmatrix, iteration); // booster is serialized and distributed among the workers from the driver
Rabit.shutdown();
```

And it seems to properly train the model :-)

3) I wasn't getting the same accuracy as I was using the new GPU plugin, which is not yet distributed and each worker was training a separate local model. The CPU updaters seem to work, so doing an initial `Rabit.init(emptyMap); XGBoost.train(data1); Rabit.shutdown()` and then a consecutive distributed training of a different model does seem to work.

**Comment by CodingCat** [comment]
are you sure they are running in a correct way?

**Comment by mdymczyk** [comment]
@CodingCat yes, did some testing:

1) as a baseline I ran a single node training using all the data I have and using only the `train` method with 50 iterations
2) then I ran a single node training on half the data and used the whole data for validation, also only `train` method with 50 iterations - the accuracy, as expected, went down
3) I ran a distributed version with 2 nodes, data split evenly half half, created a booster with `train` method with 0 iterations and then called on each mode `update` 50 times, each node was getting only half of the data (made sure of that) - the resulting accuracy is almost identical to 1).

This was with CPU updaters, when trying out GPU updaters it didn't work, each node was training a separate model using local data and no reduction took place. It is on the GPU plugin roadmap, though.
